### ISSUE TITLE
Is the dataset the runtime efficiency (MFU) bottleneck?

### ISSUE THREAD

**Issue by carbonscott** [carbonscott/peaknet#3]
With dummy data, `mfu_per_iteration = 0.32` on average. 
With safetensor data, `mfu_per_iteration = 0.14` on average, which is more than 2X slower.  I suspect the original image size being 2048x2048 is a big bottleneck.

Need a curve "mfu vs image size".

**Comment by carbonscott** [comment]
Experiment results: https://docs.google.com/spreadsheets/d/1qPh7ahleJyavWeWcDyCFyGSPPWjvLUzEmSv2ELoHNuQ/edit?gid=0#gid=0

**Comment by carbonscott** [comment]
Delay the data transformation until just before the model’s forward pass to perform the transformation all at once on the GPU.

I am achieving 

```
 iteration=151   segment=0-19200  0.342
 iteration=152   segment=0-19200  0.343
 iteration=153   segment=0-19200  0.387
 iteration=154   segment=0-19200  0.487
 iteration=155   segment=0-19200  0.489
 iteration=156   segment=0-19200  0.489
 iteration=157   segment=19200-38400  0.330
 iteration=158   segment=19200-38400  0.342
 iteration=159   segment=19200-38400  0.410
 iteration=160   segment=19200-38400  0.488
 iteration=161   segment=19200-38400  0.488
 iteration=162   segment=19200-38400  0.488
 iteration=163   segment=38400-57600  0.339
 iteration=164   segment=38400-57600  0.339
 iteration=165   segment=38400-57600  0.407
 iteration=166   segment=38400-57600  0.488
 iteration=167   segment=38400-57600  0.488
 iteration=168   segment=38400-57600  0.488
 iteration=169   segment=57600-76800  0.329
 iteration=170   segment=57600-76800  0.340
 iteration=171   segment=57600-76800  0.407
 iteration=172   segment=57600-76800  0.489
 iteration=173   segment=57600-76800  0.486
 iteration=174   segment=57600-76800  0.488
 iteration=175   segment=76800-80000  0.421
```

It did triggered cuda malloc retry: 

```
--> cuda max reserved memory = 33.0117
--> max reserved percentage = 83.8 %

--> cuda max memory allocated = 28.8766
--> max allocated percentage = 73.3 %

--> peak active memory = 28.8766
--> peak active memory 73.3 %

cudaMalloc retries = 1
cuda OOM = 0

--> Recommend decreasing batch size...cuda retries can greatly degrade perf!
```

I might need to lower the batch size accordingly.

**Comment by carbonscott** [comment]
Lowered the batch size, avoided using `torch.cat` that causes memory duplication.  I'm getting

```
 iteration=1   segment=0-18240  0.122
 iteration=2   segment=0-18240  0.499
 iteration=3   segment=0-18240  0.510
 iteration=4   segment=0-18240  0.508
 iteration=5   segment=0-18240  0.509
 iteration=6   segment=0-18240  0.509
 iteration=7   segment=18240-36480  0.373
 iteration=8   segment=18240-36480  0.384
 iteration=9   segment=18240-36480  0.410
 iteration=10   segment=18240-36480  0.509
 iteration=11   segment=18240-36480  0.509
 iteration=12   segment=18240-36480  0.510
 iteration=13   segment=36480-54720  0.375
 iteration=14   segment=36480-54720  0.385
 iteration=15   segment=36480-54720  0.410
 iteration=16   segment=36480-54720  0.508
 iteration=17   segment=36480-54720  0.508
 iteration=18   segment=36480-54720  0.509
 iteration=19   segment=54720-72960  0.377
 iteration=20   segment=54720-72960  0.393
 iteration=21   segment=54720-72960  0.405
 iteration=22   segment=54720-72960  0.510
 iteration=23   segment=54720-72960  0.509
 iteration=24   segment=54720-72960  0.508
 iteration=25   segment=72960-76000  0.459
 iteration=26   segment=0-18240  0.382
 iteration=27   segment=0-18240  0.388
 iteration=28   segment=0-18240  0.412
 iteration=29   segment=0-18240  0.511
 iteration=30   segment=0-18240  0.509
 iteration=31   segment=0-18240  0.509
 iteration=32   segment=18240-36480  0.379
 iteration=33   segment=18240-36480  0.382
 iteration=34   segment=18240-36480  0.404
```

**Comment by carbonscott** [comment]
On real data, I need to have `pre_transforms`that basically make sure all input sizes are normalized to the same one.  I'm getting

```
 iteration=1   segment=0-4800   mfu_per_iteration=0.181
 iteration=2   segment=0-4800   mfu_per_iteration=0.404
 iteration=3   segment=0-4800   mfu_per_iteration=0.409
 iteration=4   segment=0-4800   mfu_per_iteration=0.413
 iteration=5   segment=0-4800   mfu_per_iteration=0.407
 iteration=6   segment=0-4800   mfu_per_iteration=0.452
 iteration=7   segment=4800-9600   mfu_per_iteration=0.348
 iteration=8   segment=4800-9600   mfu_per_iteration=0.349
 iteration=9   segment=4800-9600   mfu_per_iteration=0.349
 iteration=10   segment=4800-9600   mfu_per_iteration=0.358
 iteration=11   segment=4800-9600   mfu_per_iteration=0.340
 iteration=12   segment=4800-9600   mfu_per_iteration=0.409
 iteration=13   segment=9600-11920   mfu_per_iteration=0.344
 iteration=14   segment=9600-11920   mfu_per_iteration=0.345
 iteration=15   segment=9600-11920   mfu_per_iteration=0.436
```

without cudamalloc retries.

```
--> cuda max reserved memory = 35.8105
--> max reserved percentage = 90.9 %

--> cuda max memory allocated = 28.8773
--> max allocated percentage = 73.3 %

--> peak active memory = 28.8773
--> peak active memory 73.3 %

cudaMalloc retries = 0
cuda OOM = 0
```
